colab link: https://colab.research.google.com/github/pedrobslima/projeto-speedrun/blob/main/projeto-speedrun.ipynb

Videogame database: https://rawg.io/

web crawling python: https://www.zenrows.com/blog/python-web-scraping-library#requests

how long to beat: https://howlongtobeat.com/

In [ ]:
from google.colab import userdata
API_KEY = userdata.get('RAWG_API_KEY')

In [ ]:
!pip install howlongtobeatpy

  Preparing metadata (setup.py) ... done
  Created wheel for howlongtobeatpy: filename=howlongtobeatpy-1.0.5-py3-none-any.whl size=10673 sha256=62c84d245a80c61b4026d5f261638a90210630b372cf8dc6b3e68fdc5167ceb1
  Stored in directory: /root/.cache/pip/wheels/ab/5c/a9/a7ed038bf3e63e09334c66de47707badb6d61bd0d8bd17cbd6
Successfully built howlongtobeatpy


In [ ]:
!pip install srcomapi
import srcomapi, srcomapi.datatypes as dt

In [ ]:
import requests
from http.client import responses
#from os.path import dirname
#from os import environ
#import inspect
import json
import gzip
import pandas as pd
import numpy as np
from time import sleep
from howlongtobeatpy import HowLongToBeat

In [ ]:
api = srcomapi.SpeedrunCom()
api.debug = 1

### Searching for a game

In [ ]:
# It's recommended to cache the game ID and use it for future requests.
# Data is cached for the current session by classname/id so future
# requests for the same game are instantaneous.
game = api.search(srcomapi.datatypes.Game, {"name": "super mario sunshine"})[0]

https://www.speedrun.com/api/v1/games


In [ ]:
api.search(srcomapi.datatypes.Game, {"name": "super mario sunshine"})

https://www.speedrun.com/api/v1/games


[<Game "Super Mario Sunshine">,
 <Game "Super Mario Sunshine 64">,
 <Game "Super Mario Sunshine Co-op">,
 <Game "Super Mario Sunshine Arcade">,
 <Game "Super Mario Sunshine Arcade 2">,
 <Game "Super Mario Sunshine Category Extensions">]

In [ ]:
game.data

{'id': 'v1pxjz68',
 'names': {'international': 'Super Mario Sunshine',
  'japanese': 'スーパーマリオサンシャイン',
  'twitch': 'Super Mario Sunshine'},
 'boostReceived': 6,
 'boostDistinctDonors': 5,
 'abbreviation': 'sms',
 'weblink': 'https://www.speedrun.com/sms',
 'discord': 'https://discord.gg/jX38aXCv9y',
 'released': 2002,
 'release-date': '2002-07-19',
 'ruleset': {'show-milliseconds': True,
  'require-verification': True,
  'require-video': False,
  'run-times': ['realtime'],
  'default-time': 'realtime',
  'emulators-allowed': True},
 'romhack': False,
 'gametypes': [],
 'platforms': [<Platform "GameCube">, <Platform "Wii">, <Platform "Switch">],
 'regions': [<Region "USA / NTSC">,
  <Region "EUR / PAL">,
  <Region "JPN / NTSC">,
  <Region "KOR / NTSC">],
 'genres': [<Genre "Platformer">, <Genre "3D Platformer">],
 'engines': [],
 'developers': [<Developer "Nintendo">],
 'publishers': [<Publisher "Nintendo">],
 'moderators': {'98r36381': 'super-moderator',
  'kj9k3lr8': 'moderator',
  'v8

### Getting the current world record for a game category

In [ ]:
game.categories
#[<Category "Any%">, ...]

[<Category "Any%">,
 <Category "120 Shines">,
 <Category "96 Shines">,
 <Category "79 Shines">,
 <Category "All Episodes">,
 <Category "Individual World">,
 <Category "Any% Hoverless">,
 <Category "All Shines">,
 <Category "Full Completion">,
 <Category "All Blue Coins">,
 <Category "20 Shines">]

In [ ]:
game.categories[0].type

'per-game'

In [ ]:
api.search(dt.Game, {"name": "Ultrakill"})

https://www.speedrun.com/api/v1/games


[<Game "ULTRAKILL">, <Game "ULTRAKILL - Category Extensions">]

In [ ]:
ultrakill = api.search(srcomapi.datatypes.Game, {"name": "Ultrakill"})[0]

https://www.speedrun.com/api/v1/games


In [ ]:
ultrakill.categories[9].variables[1].values["values"]['5legp8z1']

{'label': 'Prelude',
 'rules': 'Play from level 0-1 until the mission complete screen of 0-5.',
 'flags': {'miscellaneous': False}}

In [ ]:
ultrakill.categories[5].data

{'id': 'mke1z682',
 'name': 'Early Access NoMo',
 'weblink': 'https://www.speedrun.com/ultrakill#early-access-nomo',
 'type': 'per-game',
 'rules': '>Please make sure to also read the Game Rules.\n\nPlay on Early Access version "EA (12/20/23)" or later.\n\nDisable Enemy Spawns and Keep Cheats Activated must be enabled. No other cheats may be enabled throughout the entirety of the run.\n\nCan be played on any single difficulty level. Must only played on one difficulty.\n\nTime starts on entering the 1st level.\n\nStarting in 0-1, the other levels will automatically follow. However, after defeating Gabriel in both 3-2 and 6-2 and clicking past the mission complete screen, you may exit to the main menu and select 4-1 and 7-1 respectively to save RTA time, clicks, and sanity. (this does not affect IGT)\n\nTaking a secret exit means you must complete the secret level.\n\nTaking an exit to a level that loops back means you must complete the main level again.\n\nGetting a false level finish m

In [ ]:
ultrakill.categories[5].variables[0].data

{'id': '38djvx48',
 'name': 'Patch',
 'category': None,
 'scope': {'type': 'global'},
 'mandatory': True,
 'user-defined': False,
 'obsoletes': True,
 'values': {'_note': '`choices` is deprecated, please use `values` instead',
  'choices': {'192zmn4q': '-- EA (4/15/24) --',
   'qoxy5o4q': 'EA (12/20/23)',
   'qznn69kq': 'EA (4/27/23)',
   'qox8k4gq': 'EA (3/16/23)',
   '814zygel': 'EA (8/16/22)',
   'z19o5rk1': 'EA (2/16/22)',
   'p12kgy41': 'EA (10/28/21)',
   '4qy9godl': 'EA (8/13/21)',
   '81p27wkl': 'EA (6/23/21)',
   'gq7w70y1': 'EA (4/13/21)',
   '81wp8wmq': 'EA (12/19/20)',
   '9qjn9e01': 'EA (11/3/20)',
   '21gd2y81': 'EA (10/13/20)',
   'z19gpzyl': 'EA (9/13/20)',
   'zqowzn2l': 'EA',
   'jqzk3r8q': '1.05',
   '21go6ynq': '1.04',
   '81wrdjml': '1.03',
   '5lmw4e8q': '1.02'},
  'values': {'192zmn4q': {'label': '-- EA (4/15/24) --'},
   'qoxy5o4q': {'label': 'EA (12/20/23)'},
   'qznn69kq': {'label': 'EA (4/27/23)'},
   'qox8k4gq': {'label': 'EA (3/16/23)'},
   '814zygel': {'la

In [ ]:
ultrakill.categories[5].records[0].runs

[{'place': 1, 'run': <Run "ULTRAKILL"/"Early Access NoMo"/yoo5q15y 384.36>},
 {'place': 2, 'run': <Run "ULTRAKILL"/"Early Access NoMo"/yllxx1ky 415.317>},
 {'place': 3, 'run': <Run "ULTRAKILL"/"Early Access NoMo"/yo0llpjy 416.152>}]

In [ ]:
ultrakill.categories[5].records[0].runs[2]

{'place': 3, 'run': <Run "ULTRAKILL"/"Early Access NoMo"/yo0llpjy 416.152>}

In [ ]:
_[0].records[0].runs
#[{'run': <Run <Game "Super Mario Sunshine">/<Category "Any%">/9mr570dy 4498>, 'place': 1}, ...]

[{'place': 1, 'run': <Run "Super Mario Sunshine"/"Any%"/znv51pvm 4388>},
 {'place': 2, 'run': <Run "Super Mario Sunshine"/"Any%"/y4gdpgny 4391>},
 {'place': 3, 'run': <Run "Super Mario Sunshine"/"Any%"/y46o31nz 4397>},
 {'place': 3, 'run': <Run "Super Mario Sunshine"/"Any%"/y4ojejqz 4397>}]

In [ ]:
game.records[0].runs

[{'place': 1, 'run': <Run "Super Mario Sunshine"/"Any%"/znv51pvm 4388>},
 {'place': 2, 'run': <Run "Super Mario Sunshine"/"Any%"/y4gdpgny 4391>},
 {'place': 3, 'run': <Run "Super Mario Sunshine"/"Any%"/y46o31nz 4397>},
 {'place': 3, 'run': <Run "Super Mario Sunshine"/"Any%"/y4ojejqz 4397>}]

In [ ]:
_[0]["run"].players
#{'primary_t': 4498, ...}
# primary_t is the time in seconds

https://www.speedrun.com/api/v1/users/qj2y1r68


[<User "JJsrl">]

### Getting a dict containing all runs from a game

In [ ]:
ultrakill_runs = {}
for category in ultrakill.categories:
  if not category.name in ultrakill_runs:
    ultrakill_runs[category.name] = {}
  if category.type == 'per-level':
    for level in ultrakill.levels:
      ultrakill_runs[category.name][level.name] = dt.Leaderboard(api, data=api.get("leaderboards/{}/level/{}/{}?embed=variables".format(ultrakill.id, level.id, category.id)))
  else:
    ultrakill_runs[category.name] = dt.Leaderboard(api, data=api.get("leaderboards/{}/category/{}?embed=variables".format(ultrakill.id, category.id)))

https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/29vl6lqw/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/xd4y1yqd/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/xd0212m9/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/rw6l1lpw/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/n93n1y2w/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/29v5x2xd/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/xd4rzx89/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/o9xx7q79/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/495m5g4w/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/rdqn7rgd/5dw44w5k?embed=variables
https://www.speedrun.com/api/v1/leaderboards/369p3p81/level/

In [ ]:
ultrakill_runs['Any%']['0-1: Into The Fire'].runs

[{'place': 1, 'run': <Run "ULTRAKILL"/"Any%"/z0nl29om 2.958>},
 {'place': 2, 'run': <Run "ULTRAKILL"/"Any%"/mee35p9m 3.848>},
 {'place': 3, 'run': <Run "ULTRAKILL"/"Any%"/y27n8r7z 3.97>},
 {'place': 4, 'run': <Run "ULTRAKILL"/"Any%"/mewek93m 4.083>},
 {'place': 5, 'run': <Run "ULTRAKILL"/"Any%"/ylrpxvky 4.278>},
 {'place': 6, 'run': <Run "ULTRAKILL"/"Any%"/meelpl8m 4.355>},
 {'place': 7, 'run': <Run "ULTRAKILL"/"Any%"/yln5d03z 5.067>},
 {'place': 8, 'run': <Run "ULTRAKILL"/"Any%"/m37jjgwz 5.239>},
 {'place': 9, 'run': <Run "ULTRAKILL"/"Any%"/z5eeqwgz 5.687>},
 {'place': 10, 'run': <Run "ULTRAKILL"/"Any%"/ydedvqxm 6.009>},
 {'place': 11, 'run': <Run "ULTRAKILL"/"Any%"/yl61rv3y 6.152>},
 {'place': 12, 'run': <Run "ULTRAKILL"/"Any%"/z19qj9wm 6.257>},
 {'place': 13, 'run': <Run "ULTRAKILL"/"Any%"/ydjj7l0y 6.639>},
 {'place': 14, 'run': <Run "ULTRAKILL"/"Any%"/yl1jq1xm 6.942>},
 {'place': 15, 'run': <Run "ULTRAKILL"/"Any%"/zxnge65m 6.951>},
 {'place': 16, 'run': <Run "ULTRAKILL"/"Any%"/y969

In [ ]:
ultrakill_runs['Early Access Any%'].runs

[{'place': 1, 'run': <Run "ULTRAKILL"/"Early Access Any%"/me0ngxqz 761.644>},
 {'place': 2, 'run': <Run "ULTRAKILL"/"Early Access Any%"/m301d3wm 801.023>},
 {'place': 3, 'run': <Run "ULTRAKILL"/"Early Access Any%"/zq805x5z 802.493>},
 {'place': 4, 'run': <Run "ULTRAKILL"/"Early Access Any%"/zq8v785z 856.651>},
 {'place': 5, 'run': <Run "ULTRAKILL"/"Early Access Any%"/ywn9wp9y 954.003>},
 {'place': 6, 'run': <Run "ULTRAKILL"/"Early Access Any%"/z0xq23oz 960.719>},
 {'place': 7, 'run': <Run "ULTRAKILL"/"Early Access Any%"/mr2vn34y 1027.568>},
 {'place': 8, 'run': <Run "ULTRAKILL"/"Early Access Any%"/yvlq24xz 1087.478>},
 {'place': 9, 'run': <Run "ULTRAKILL"/"Early Access Any%"/z192v7gm 1132.77>},
 {'place': 10, 'run': <Run "ULTRAKILL"/"Early Access Any%"/m3k826wy 1134.373>},
 {'place': 11, 'run': <Run "ULTRAKILL"/"Early Access Any%"/y6q4416z 1284.952>},
 {'place': 12, 'run': <Run "ULTRAKILL"/"Early Access Any%"/zp6w828m 1334.764>},
 {'place': 13, 'run': <Run "ULTRAKILL"/"Early Access Any

In [ ]:
# info completa do jogo https://www.speedrun.com/api/v1/games/369p3p81?embed=categories.variables,levels,genres
ultratest = requests.get("https://www.speedrun.com/api/v1/games/369p3p81?embed=categories.variables,levels,genres").json()["data"]

In [ ]:
test = requests.get("https://www.speedrun.com/api/v1/games?genre=qdnqyk28").json()["data"]

In [ ]:
ultratest

{'id': '369p3p81',
 'names': {'international': 'ULTRAKILL',
  'japanese': None,
  'twitch': 'ULTRAKILL'},
 'boostReceived': 30,
 'boostDistinctDonors': 22,
 'abbreviation': 'ultrakill',
 'weblink': 'https://www.speedrun.com/ultrakill',
 'discord': 'https://discord.gg/KgkVuXxQgB',
 'released': 2019,
 'release-date': '2019-08-30',
 'ruleset': {'show-milliseconds': True,
  'require-verification': True,
  'require-video': True,
  'run-times': ['realtime', 'ingame'],
  'default-time': 'ingame',
  'emulators-allowed': False},
 'romhack': False,
 'gametypes': [],
 'platforms': ['8gej2n93'],
 'regions': [],
 'genres': {'data': [{'id': 'm65jk2k0',
    'name': 'FPS',
    'links': [{'rel': 'self',
      'uri': 'https://www.speedrun.com/api/v1/genres/m65jk2k0'},
     {'rel': 'games',
      'uri': 'https://www.speedrun.com/api/v1/games?genre=m65jk2k0'}]},
   {'id': '94n4pn0w',
    'name': 'Action',
    'links': [{'rel': 'self',
      'uri': 'https://www.speedrun.com/api/v1/genres/94n4pn0w'},
     {

In [ ]:
dfgames = pd.DataFrame(columns = ['game_id','name', 'release-date', 'genres',     'levels',    'categories'])#, 'sub-catg'])
                                  # str      str       date/str     list[str]   None|list[str]   list[str]   None|list[str]
dfruns = pd.DataFrame(columns = ['run_id', 'game_id', 'user_id', 'time', 'upload-date', 'level',  'category', 'status'])#,   'sub-catg'])
                                  # str       str        str      str      date/str     None|str      str        str         None|list[str]
dfusers = pd.DataFrame(columns = ['user_id', 'username', 'location', 'region', 'signup-date'])
                                  # str         str       None|str   None|str    date/str

# metrics.at['Decision Tree','Acurácia'] = accuracy_score(y_test, y_predDT)

In [ ]:
response = requests.get(f"https://api.rawg.io/api/games?genres=platformer&key={API_KEY}&ordering=-metacritic&page=18&page_size=40")

In [ ]:
test = response.json()["results"]

for i in range(len(test)):
  print(test[i]["name"])
  print(test[i]["tba"], type(test[i]["tba"]))

advantica
False <class 'bool'>
Cubed The Demo
False <class 'bool'>
Night Keeper
False <class 'bool'>
Impossible Castle: BAM Boys 2
False <class 'bool'>
BloodTower
False <class 'bool'>
Transmit Thesis
False <class 'bool'>
The Abyss (II BruCeY II)
False <class 'bool'>
Anti Jellyfish Force
False <class 'bool'>
Bouncing Billy
False <class 'bool'>
Sanctuary (Nudders)
False <class 'bool'>
Trumpet
False <class 'bool'>
Tiny Parkour
False <class 'bool'>
Loco The Firefox Free HD v1.0R
False <class 'bool'>
Mushroom High
False <class 'bool'>
SkyRider Prototype
False <class 'bool'>
Stans Night out
False <class 'bool'>
Some Happier Star
False <class 'bool'>
YOU CAN ONLY DIE 80 TIMES
False <class 'bool'>
CYANiDE
False <class 'bool'>
BlipBloop
False <class 'bool'>
Transmute
False <class 'bool'>
Penguin Savior
False <class 'bool'>
Rock Pit
False <class 'bool'>
Train Jumper
False <class 'bool'>
Climber (Quantinium)
False <class 'bool'>
Stock Graphics Jungle (beta)
False <class 'bool'>
Beatrice (shonenki

In [ ]:
allData

[{'name': 'Super Mario Galaxy',
  'release date': '2007-11-01',
  'platforms': ['Wii'],
  'genres': ['Platformer']},
 {'name': 'Super Mario Galaxy 2',
  'release date': '2010-05-23',
  'platforms': ['Wii U', 'Wii'],
  'genres': ['Platformer']},
 {'name': 'Super Mario Odyssey',
  'release date': '2017-10-27',
  'platforms': ['Nintendo Switch'],
  'genres': ['Platformer', 'Arcade']},
 {'name': 'Super Mario 64',
  'release date': '1996-06-23',
  'platforms': ['Nintendo 64'],
  'genres': ['Platformer']},
 {'name': 'Super Mario Advance 4: Super Mario Bros. 3',
  'release date': '2003-07-11',
  'platforms': ['Game Boy Advance'],
  'genres': ['Platformer']},
 {'name': 'Jet Set Radio',
  'release date': '2000-06-29',
  'platforms': ['PC',
   'iOS',
   'Android',
   'Xbox 360',
   'PlayStation 3',
   'PS Vita',
   'Game Boy Advance',
   'Dreamcast'],
  'genres': ['Platformer', 'Action', 'Sports']},
 {'name': 'Super Mario 3D World',
  'release date': '2013-11-21',
  'platforms': ['Nintendo Switc

In [ ]:
# requests.get("https://www.speedrun.com/api/v1/games/369p3p81?embed=categories,variables,levels,genres").json()["data"]
# https://steamdb.info/tag/1625/?min_reviews=500&displayOnly=Game&sort=peak_desc
#url = "https://www.speedrun.com/api/v1/games?genre=qdnqyk28&max=200&offset=0&embed=genres"
#url = "https://www.speedrun.com/api/v1/games?_bulk=yes&genre=qdnqyk28&max=1000&offset=0"
url = f"https://api.rawg.io/api/games?exclude_additions=true&genres=83&key={API_KEY}&page_size=40"
check = True
page = 0
allData = []
while check:
  response = requests.get(url)  #api.get(BASE_URL + str(i))
  page += 1
  print(url, "page:", page, "status_code:", response.status_code)

  if(response.status_code != 200):
    tries = 0
    while(response.status_code != 200 and tries < 5):
      print(f"dole {tries+1} ({response.status_code})")
      sleep(5)
      response = requests.get(url)
      tries += 1
  #if(response.status_code == 420):
  #  print("Limit reached. Sleeping for 2min.")
  #  sleep(120)
  #else:
  #  data = response.json()["data"]
  #  pagination = response.json()["pagination"]
  #pagination = response.json()["pagination"]

  if(response.status_code == 200):
    data = response.json()["results"]
    next = response.json()["next"]
    for game in data:
      if(("2024-05-13" >= str(game["released"])) and (int(game["ratings_count"]) > 1)):
        genres = []
        for g in game["genres"]:
          genres.append(g["name"])
        platforms = []
        if(game["platforms"] != None):
          for p in game["platforms"]:
            platforms.append(p["platform"]["name"])
        allData.append({"name": game["name"], "release date": game["released"], "platforms": platforms, "genres": genres})

    if(next == "null" or page==250):
      check = False
    else:
      url = next
    #print(count)
  else:
    print("error:", response.status_code)
    check = False

    #for game in data:
     # allData.append(game)
  '''#genres
      grs = []
      for gdata in game["genres"]["data"]:
        grs.append(gdata["name"])
        #print(gdata["id"], gdata["name"])
      #levels
      if("levels" not in game.keys()):
        lvls = pd.NaN
      else:
        lvls = []
        for ldata in game["levels"]["data"]:
          lvls.append(ldata["name"])
          #print(ldata["id"], ldata["name"])
      #categories
      if("categories" not in game.keys()):
        ctgs = pd.NaN
      else:
        ctgs = []
        for cdata in game["categories"]["data"]:
          ctgs.append(cdata["name"])
          #print(cdata["id"], cdata["name"])
      #sub-catg
      #if("variables" not in game.keys()):
      #  vars = pd.NaN
      #else:
      #  vars = []
      #  for vardata in cdata["variables"]["data"]:
      #    print('\t', vardata["id"], vardata["name"])
      #    for idv in vardata["values"]["values"]:
      #      print('\t\t', idv, vardata["values"]["values"][idv]["label"])

      #['game_id','name', 'release-date', 'genres',     'levels',    'categories', 'sub-catg']
      dfgames.loc[len(dfgames.index)] = [game["id"], game["names"]["international"], game["release-date"], grs, lvls, ctgs]'''
  '''
  count += int(pagination["size"])
  if(int(pagination["size"]) >= 200):
    if(pagination["links"][-1]["rel"] == 'next'):
      url = pagination["links"][-1]["uri"]
  else:
    check = False'''


https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page_size=40 page: 1 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=2&page_size=40 page: 2 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=3&page_size=40 page: 3 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=4&page_size=40 page: 4 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=5&page_size=40 page: 5 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=6&page_size=40 page: 6 status_code: 200
https://api.rawg.io/api/games?exclude_additions=true&genres=83&key=c460f9210afb429aa4e685eb2c05ab47&page=7&page_size=40 page: 7 status_co

In [ ]:
len(allData)

1227

In [ ]:
for i in range(len(allData)):
  if(allData[i]["name"] == "Super Mombo Quest"):
    print(allData[i])

{'name': 'Super Mombo Quest', 'release date': '2021-11-03', 'platforms': ['PC', 'iOS', 'Android'], 'genres': ['Indie', 'Platformer', 'Adventure', 'Action']}


In [ ]:
for i in range(len(allData)):
  if(allData[i]["name"] == "Super Mombo Quest"):
    print(allData[i])

In [ ]:
results = HowLongToBeat().search("It Takes Two")

In [ ]:
results

In [ ]:
print(results[0].game_name, results[0].main_story, results[0].main_extra, results[0].completionist)

It Takes Two 12.66 14.0 15.87


In [ ]:
results[2].similarity

0.5925925925925926

In [ ]:
results_list = await HowLongToBeat().async_search("Metroid ii return of samus")
results_list[0].similarity

0.8301886792452831

In [ ]:
test = "metroid II - Return of Samus"

In [ ]:
test = test.capitalize()
print(test)

Metroid ii - return of samus


In [ ]:
test.replace("-", "")

'Metroid II  Return of Samus'

In [ ]:
results_list[0].release_world

1993

In [ ]:
from difflib import SequenceMatcher
print(SequenceMatcher(None, "Mega Man X", "Mega Man X (1993)").ratio())

0.7407407407407407


In [ ]:
for game in allData:
  game["name"] = game["name"].replace("’", "'")

In [ ]:
for i in range(len(allData)):
  if(allData[i]["release date"] == None):
    print(i, allData[i])

In [ ]:
data = "2024-05-12"
print("2024-05-13" >= data)

True


6 jogos sem datas de lançamento, 3 atualizados manualmente, 3 foram excluídos

In [ ]:
for game in allData:
  if(game["name"][-6:] == str("("+game["release date"][:4]+")")):
    results_list = await HowLongToBeat().async_search(game["name"][:-6])
    best = 0
    for r in results_list:
      if(r.similarity > best and r.release_world == game["release date"][:4]):
        best = r.similarity
        best_element = r
  else:
    results_list = await HowLongToBeat().async_search(game["name"])

  if(results_list is None or len(results_list) == 0):
    replace = game["name"].capitalize()
    for token in [":", "-", ".", "!", "(", ")", "?"]:
      replace = replace.replace(token, "")
    results_list = await HowLongToBeat().async_search(replace)

  if results_list is not None and len(results_list) > 0:
    best_element = max(results_list, key=lambda element: element.similarity)

    game["main story"] = best_element.main_story
    game["main + extra"] = best_element.main_extra
    game["100%"] = best_element.completionist

  elif(len(results_list) == 0):
    print("Não achou", game["name"])

Não achou SEGA Mega Drive and Genesis Classics
Não achou BIT.TRIP Presents... Runner2: Future Legend of Rhythm Alien
Não achou Inmost
Não achou Another World – 20th Anniversary Edition
Não achou Oniken: Unstoppable Edition
Não achou Life Goes On: Done to Death
Não achou Disney's DuckTales
Não achou Synergy ( Ske)
Não achou Super Monsters'n Girls
Não achou SpongeBob: Truth-Sq.
Não achou Catto Boi: Journey to Catto Land
Não achou Sonic 3 and Knuckles
Não achou Sanabi
Não achou Runner2
Não achou WALL•E
Não achou Lewd Leaf Land - Maple Tea Ecstasy
Não achou The Dishwasher: VS
Não achou Eversleep: Lunistice Gaiden
Não achou B͞҉ŗeak the̶̢ ͠҉G̵ame
Não achou Disney's TaleSpin (TurboGrafx-16)
Não achou SONIC 4 Episode II
Não achou Point Taker
Não achou Shinobi (1988)
Não achou Dangerous Dave Pack
Não achou Takahashi Meijin no Bugutte Honey
Não achou MegaMan Zero
Não achou Prehistorik 1+2
Não achou Night at the Museum 2
Não achou The Magical Quest (Starring Mickey Mouse)
Não achou Tomb Raider: 

250 jogos não encontrados por busca automatizada sem tratamento de nomes

136 jogos não encontrados por busca automatizada com tratamento de nomes

In [ ]:
df = pd.DataFrame.from_dict(allData)

In [ ]:
df.head()

name release date  \
0              Limbo   2010-07-21   
1           Terraria   2011-05-16   
2      Hollow Knight   2017-02-23   
3  Little Nightmares   2017-04-27   
4     Super Meat Boy   2010-10-20   

                                           platforms  \
0  [PC, Xbox One, PlayStation 4, Nintendo Switch,...   
1  [PC, Xbox One, PlayStation 4, Nintendo Switch,...   
2  [PC, Xbox One, PlayStation 4, Nintendo Switch,...   
3     [PC, Xbox One, PlayStation 4, Nintendo Switch]   
4  [PC, PlayStation 4, Nintendo Switch, Android, ...   

                                genres  main story  main + extra    100%  
0  [Platformer, Indie, Action, Puzzle]        3.58          4.17    6.64  
1          [Platformer, Indie, Action]       51.91         91.53  194.94  
2          [Platformer, Indie, Action]       26.90         41.52   63.59  
3                 [Platformer, Action]        3.55          4.97    8.32  
4                  [Platformer, Indie]        9.46         18.52   43.91

In [ ]:
df[df['main story'].isna()]

name release date  \
5                  SEGA Mega Drive and Genesis Classics   2010-06-01   
83    BIT.TRIP Presents... Runner2: Future Legend of...   2013-02-25   
136                                              Inmost   2019-10-11   
147            Another World – 20th Anniversary Edition   2011-09-22   
183                         Oniken: Unstoppable Edition   2014-02-05   
...                                                 ...          ...   
1222  Dora Diginoid: Metroidvania game (made in Godo...   2022-06-27   
1223                                             Exit84   2018-03-05   
1224                            Lonk's Greedy Adventure   2017-12-04   
1225                                 Jimbo (Agustin D.)   2022-02-25   
1226                         Байки черных котов: Начало   2020-12-09   

                                              platforms  \
5     [PC, Xbox One, PlayStation 4, Nintendo Switch,...   
83    [PC, PlayStation 4, macOS, Linux, Xbox 360, Pl...   
136                   [PC, Nintendo Switch, iOS, macOS]   
147   [PC, PlayStation 4, iOS, macOS, Linux, PlaySta...   
183                [PC, PlayStation 4, Nintendo Switch]   
...                                                 ...   
1222                                               [PC]   
1223                            [PC, macOS, Linux, Web]   
1224                                               [PC]   
1225                                              [Web]   
1226                                               [PC]   

                                              genres  main story  \
5     [Adventure, Action, RPG, Fighting, Platformer]         NaN   
83    [Adventure, Action, Arcade, Indie, Platformer]         NaN   
136                   [Indie, Platformer, Adventure]         NaN   
147          [Arcade, Platformer, Adventure, Action]         NaN   
183                      [Platformer, Indie, Action]         NaN   
...                                              ...         ...   
1222                [Platformer, Shooter, Adventure]         NaN   
1223                                    [Platformer]         NaN   
1224                               [Platformer, RPG]         NaN   
1225                         [Platformer, Adventure]         NaN   
1226                         [Platformer, Adventure]         NaN   

      main + extra  100%  
5              NaN   NaN  
83             NaN   NaN  
136            NaN   NaN  
147            NaN   NaN  
183            NaN   NaN  
...            ...   ...  
1222           NaN   NaN  
1223           NaN   NaN  
1224           NaN   NaN  
1225           NaN   NaN  
1226           NaN   NaN  

[136 rows x 7 columns]

In [ ]:
df[df["name"] == "Lonk's Greedy Adventure"].index[0]

1224

In [ ]:
df["main story"].dtypes

dtype('float64')



---
Adding manually


In [ ]:
name = "WALL•E"
namehl2b = "WALL-E"

In [ ]:
results = await HowLongToBeat().async_search(namehl2b)
res = results[0]
res

In [ ]:
idx = df[df["name"] == name].index[0]
df.at[idx, 'main story'] = res.main_story
df.at[idx, 'main + extra'] = res.main_extra
df.at[idx, '100%'] = res.completionist

In [ ]:
df[df["name"] == name]

name release date platforms                                  genres  \
434  Sanabi   2023-11-09      [PC]  [Indie, Platformer, Adventure, Action]   

     main story  main + extra   100%  
434        8.46         10.26  14.97

In [ ]:
df.to_csv(r"/content/gamesdf.csv", index=False)

NameError: name 'df' is not defined

In [ ]:
teste = None
if(teste is None or len(teste)>0):
  print("a")

a


In [ ]:
print('d' in test.values())

True




---



In [ ]:
!pip install HowLongToBeat-Python-API
from hltbapi import HtmlScraper

In [ ]:
# id
print("---ID---")
print(ultratest["id"])

# name
print("---NAME---")
print(ultratest["names"]["international"])

# release date
print("---RELEASE---")
print(ultratest["release-date"])

# genres
print("---GENRES---")
for gdata in ultratest["genres"]["data"]:
  print(gdata["id"], gdata["name"])

# levels
print("---LEVELS---")
for ldata in ultratest["levels"]["data"]:
  print(ldata["id"], ldata["name"])

#for i in range(len(ultratest["levels"]["data"])):
#  print(ultratest["levels"]["data"][i]["id"], ultratest["levels"]["data"][i]["name"])

# categories and variables
print("---CATEGORIES&SUBCATG---")
for cdata in ultratest["categories"]["data"]:
  print(cdata["id"], cdata["name"])
  for vardata in cdata["variables"]["data"]:
    print('\t', vardata["id"], vardata["name"])
    for idv in vardata["values"]["values"]:
      print('\t\t', idv, vardata["values"]["values"][idv]["label"])

#for i in range(len(ultratest["categories"]["data"])):
#  print(ultratest["categories"]["data"][i]["id"], ultratest["categories"]["data"][i]["name"])
#  for j in range(len(ultratest["categories"]["data"][i]["variables"]["data"])):
#    print('\t', ultratest["categories"]["data"][i]["variables"]["data"][j]["id"], ultratest["categories"]["data"][i]["variables"]["data"][j]["name"])
#    for idv in ultratest["categories"]["data"][i]["variables"]["data"][j]["values"]["values"]:
#      print('\t\t', idv, ultratest["categories"]["data"][i]["variables"]["data"][j]["values"]["values"][idv]["label"])

---ID---
369p3p81
---NAME---
ULTRAKILL
---RELEASE---
2019-08-30
---GENRES---
m65jk2k0 FPS
94n4pn0w Action
---LEVELS---
29vl6lqw 0-1: Into The Fire
xd4y1yqd 0-2: The Meatgrinder
xd0212m9 0-3: Double Down
rw6l1lpw 0-4: A One-Machine Army
n93n1y2w 0-5: Cerberus
29v5x2xd 0-S: Something Wicked
xd4rzx89 1-1: Heart Of The Sunrise
o9xx7q79 1-2: The Burning World
495m5g4w 1-3: Halls Of Sacred Remains
rdqn7rgd 1-4: Clair De Lune
xd03yxjd 1-S: The Witless
5d7245ew 2-1: Bridgeburner
kwjvzj1w 2-2: Death At 20,000 Volts
owo36n3w 2-3: Sheer Heart Attack
xd142ryd 2-4: Court Of The Corpse King
ewp8xy49 2-S: All-Imperfect Love Song
y9m2x30d 3-1: Belly Of The Beast
5wkq6oxw 3-2: In The Flesh
y9m2rqzd P-1: Soul Survivor
n93kk87w 4-1: Slaves To Power
z98ppe7w 4-2: God Damn The Sun
rdn33rnw 4-3: A Shot In The Dark
ldyqqjjw 4-4: Clair De Soleil
gdrrrg8d 4-S: Clash of the Brandicoot
nwlv8lo9 5-1: In The Wake Of Poseidon
ywe7o1ld 5-2: Waves Of The Starless Sea
69zv13xw 5-3: Ship Of Fools
r9gop2jw 5-4: Leviatha

In [ ]:
count = 0
for category in ultrakill.categories:
  if category.type == 'per-level':
    for level in ultrakill.levels:
      count += len(ultrakill_runs[category.name][level.name].runs)
  else:
    count += len(ultrakill_runs[category.name].runs)
print(count)

10583


In [ ]:
ultrakill.variables

https://www.speedrun.com/api/v1/games/369p3p81/variables


[<Variable "Patch">,
 <Variable "Act Number">,
 <Variable "Act Number P Ranks">,
 <Variable "Chapter Number">,
 <Variable "Chapter Number P Ranks">,
 <Variable "CHAPTER NUMBER NoMo">,
 <Variable "4-S contingencies">,
 <Variable "Exit 0-2">,
 <Variable "Exit 0-2 NoMo">,
 <Variable "Exit 1-1">,
 <Variable "Exit 1-1 NoMo">,
 <Variable "Exit 2-3">,
 <Variable "Exit 2-3 NoMo">,
 <Variable "Exit 3-1">,
 <Variable "Exit 3-1 NoMo">,
 <Variable "Exit 4-2">,
 <Variable "Exit 4-2 NoMo">,
 <Variable "Exit 5-1">,
 <Variable "Exit 5-1 NoMo">,
 <Variable "Exit 6-2">,
 <Variable "Exit 6-2 NoMo">,
 <Variable "Exit 7-3 (broke)">,
 <Variable "Exit 7-3">,
 <Variable "Exit 7-3 NoMo">]

In [ ]:
# modificar para pegar as informações das runs (e lidar com categorias com mais de 20k runs (e implementar uma forma de ver essas 20k runs/categoria))
def get_runs(game_id: str, category_id: str, level_id: str = None) -> int:
  if level_id:
    url = f"https://www.speedrun.com/api/v1/runs?game={game_id}&max=200&level={level_id}&category={category_id}&offset=0"
  else:
    url = f"https://www.speedrun.com/api/v1/runs?game={game_id}&max=200&category={category_id}&offset=0"

  check = True
  count = 0

  while check:
    response = requests.get(url)  #api.get(BASE_URL + str(i))
    print(url, "count:", count, "status_code:", response.status_code)
    if(response.status_code == 420):
      print("Limit reached. Sleeping for 2min.")
      sleep(120)
    else:
      pagination = response.json()["pagination"]
      #for i in range(len(data)):
        #idgames.append(data[i]['id'])
      count += int(pagination["size"])
      if(pagination["size"] >= 200):
        if(pagination["links"][-1]["rel"] == 'next'):
          url = pagination["links"][-1]["uri"]
      else:
        check = False

      print(count)
  return count

In [ ]:
ultrakill_allruns = {}
ultracount = 0
for category in ultrakill.categories:
  if category.type == 'per-level':
    for level in ultrakill.levels:
      ultracount += get_runs(ultrakill.id, category.id, level.id)
      #ultrakill_allruns[category.name][level.name] = dt.Run(api, data=api.get("runs?game={}&level={}&category={}&embed=variables".format(ultrakill.id, level.id, category.id)))
  else:
    ultracount += get_runs(ultrakill.id, category.id)
    #ultrakill_allruns[category.name] = dt.Run(api, data=api.get("runs/{}?game={}category={}&embed=variables".format(ultrakill.id, category.id)))
  print("total:", ultracount)

https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=29vl6lqw&category=5dw44w5k&offset=0 count: 0 status_code: 200
200
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=29vl6lqw&category=5dw44w5k&offset=200 count: 200 status_code: 200
400
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=29vl6lqw&category=5dw44w5k&offset=400 count: 400 status_code: 200
557
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=xd4y1yqd&category=5dw44w5k&offset=0 count: 0 status_code: 200
200
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=xd4y1yqd&category=5dw44w5k&offset=200 count: 200 status_code: 200
245
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=xd0212m9&category=5dw44w5k&offset=0 count: 0 status_code: 200
200
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=xd0212m9&category=5dw44w5k&offset=200 count: 200 status_code: 200
287
https://www.speedrun.com/api/v1/runs?game=369p3p81&max=200&level=rw

In [ ]:
url = "https://www.speedrun.com/api/v1/runs?game=ultrakill&max=200&category=wkpwoxgd&offset=0"
check = True
count = 0

while check:
  response = requests.get(url)  #api.get(BASE_URL + str(i))
  print(url, "count:", count, "status_code:", response.status_code)
  pagination = response.json()["pagination"]
  #for i in range(len(data)):
    #idgames.append(data[i]['id'])
  count += int(pagination["size"])
  if(pagination["size"] >= 200):
  #if(pagination["links"][-1]["rel"] == 'next'):
    url = pagination["links"][-1]["uri"]
    #print(count)
  else:
    check = False
print(count)

https://www.speedrun.com/api/v1/runs?game=ultrakill&max=200&category=wkpwoxgd&offset=0 count: 0 status_code: 200
61


In [ ]:
dfk = pd.read_csv("/content/runs-data.csv", sep=",")

In [ ]:
dfk.head()

ID    gameID categoryID   levelID        date  primaryTime  platform  \
0  z159r8wy  yd4ovvg1   9kv7jyok       NaN  2020-07-11      329.000  lq60nl94   
1  mke31rvm  o1y9pyk6   mke9xe9d  ldyrq3r9  2021-12-25       17.550  o7e25xew   
2  y205n39m  o1y9pyk6   mke9xe9d  ewp8q4z9  2021-12-25        7.573  o7e25xew   
3  z0o534jz  o1y9pyk6   mke9xe9d  5d7e0vvw  2021-12-22       12.420  o7e25xew   
4  yorkj2jy  o1y9pyk6   mke9xe9d  5d7e0vvw  2021-12-22       17.820  o7e25xew   

   emulated   players  examiner  \
0     False  xy5275w8  7j49e3w8   
1     False  jpmez55j  8wken9v8   
2     False  jpmez55j  8wken9v8   
3     False  jpmez55j  98rl53w8   
4     False  jpmez55j  98rl53w8   

                                              values    status statusReason  \
0  5ly7759l=21dy6xj1,e8mmmoq8=jqzdy82l,ylqkkw3l=4...  verified          NaN   
1                0nwovxdl=4qyoge3l,p854j77l=81pw5rel  verified          NaN   
2                0nwovxdl=4qyoge3l,p854j77l=81pw5rel  verified          NaN   
3                0nwovxdl=4qyoge3l,p854j77l=81pw5rel  verified          NaN   
4                0nwovxdl=5q88w7rq,p854j77l=81pw5rel  verified          NaN   

           verifiedDate  
0  2020-07-12T09:30:20Z  
1  2021-12-27T09:40:30Z  
2  2021-12-27T09:40:54Z  
3  2021-12-23T09:02:54Z  
4  2021-12-23T09:03:38Z

In [ ]:
len(dfk[dfk["gameID"]==ultrakill.id])

14481

In [ ]:
len(dfk[dfk["gameID"]=="o1y9j9v6"]) # celeste

35351

Obs.: Um dos motivos de não ter sido utilizado um dataset similar a este encontrado no kaggle, foi o tempo desde sua publicação, que já faz 6 meses. Um exemplo da diferença do mesmo banco de dados de 6 meses atrás, é com o jogo Ultrakill, de Arsi "Hakita" Patala. O jogo vêm aumentando sua comunidade como um todo, assim como sua comunidade de speedrun, possivelmente dada à jogabilidade frenética e às mecânicas do jogo, que incentivam jogadores a terminar níveis o mais rápido possível, e com pontuações altas. [Mostrar número de jogadores ao longo do tempo com steamDB] Além disso, com o jogo ainda em desenvolvimento, cada vez que novos conteúdos são adicionados, uma nova leva de jogadores é levada a jogar. Todos esses fatores fizeram com que cerca de 6000 speedruns novas fossem publicadas no speedrun.com durante este intervalo de 6 meses (antes eram 14481, agora está por volta dos 20300), lhe colocando como o 10º jogo com maior número de speedruns.

Problema principal: https://github.com/speedruncomorg/api/issues/125

### Outros


In [ ]:
api.get("games?_bulk=yes&max=1000")

https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000


[{'id': 'v1plx046',
  'names': {'international': ' Battle Mania', 'japanese': None},
  'abbreviation': '_Battle_Mania',
  'weblink': 'https://www.speedrun.com/_Battle_Mania'},
 {'id': 'o1y32j26',
  'names': {'international': ' Bike Unchained 2', 'japanese': None},
  'abbreviation': '_bike_unchained_2',
  'weblink': 'https://www.speedrun.com/_bike_unchained_2'},
 {'id': 'y65rm041',
  'names': {'international': " BIRDIE WING -Golf Girls' Story-",
   'japanese': None},
  'abbreviation': '_birdie_wing_golf_girls_story',
  'weblink': 'https://www.speedrun.com/_birdie_wing_golf_girls_story'},
 {'id': 'j1n4pkx6',
  'names': {'international': ' Body by Milk Race and Refuel',
   'japanese': None},
  'abbreviation': 'bbmrr',
  'weblink': 'https://www.speedrun.com/bbmrr'},
 {'id': 'kdkqz3xd',
  'names': {'international': ' DYNAMITE BOMB!! ', 'japanese': None},
  'abbreviation': '_DYNAMITE_BOMB_',
  'weblink': 'https://www.speedrun.com/_DYNAMITE_BOMB_'},
 {'id': '9d3wwkg1',
  'names': {'internatio

In [ ]:
response = requests.get(url="https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000&offset=0")

In [ ]:
response.json()["pagination"]["links"]

[{'rel': 'next',
  'uri': 'https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000&offset=1000'}]

In [ ]:
response.json()["pagination"]["links"][-1]["uri"]

'https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000&offset=1000'

In [ ]:
type(response.json()["data"])

list

In [ ]:
api.search(dt.Game, {"name": "Battle Mania"})

https://www.speedrun.com/api/v1/games


[<Game " Battle Mania">, <Game "Battle Mania: Daiginjo">]

# 1) Coleta de dados

## Games

In [ ]:
dfgames = pd.DataFrame(columns = ['game_id','name', 'release date',])
# metrics.at['Decision Tree','Acurácia'] = accuracy_score(y_test, y_predDT)

id list of all games

In [ ]:
idgames = []
url = "https://www.speedrun.com/api/v1/games?max=200&offset=0"
check = True

while check:
  response = requests.get(url).json()  #api.get(BASE_URL + str(i))
  data = response["data"]
  for i in range(len(data)):
    idgames.append(data[i]['id'])
  if(response["pagination"]["links"][-1]["rel"] == 'next'):
    url = response["pagination"]["links"][-1]["uri"]
  else:
    check = False

In [ ]:
len(idgames)

39991

In [ ]:
api.get("games?_bulk=yes&max=1000&offset=0")[0]

https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000&offset=0


{'id': 'v1plx046',
 'names': {'international': ' Battle Mania', 'japanese': None},
 'abbreviation': '_Battle_Mania',
 'weblink': 'https://www.speedrun.com/_Battle_Mania'}

In [ ]:
len(api.get("games?_bulk=yes&max=1000&offset=39000"))

https://www.speedrun.com/api/v1/games?_bulk=yes&max=1000&offset=39000


991

In [ ]:
user = dt.User(api, id="j59wdvnx")

In [ ]:
user.data

{'id': 'j59wdvnx',
 'names': {'international': 'rewrkk', 'japanese': None},
 'supporterAnimation': False,
 'pronouns': 'Any/All',
 'weblink': 'https://www.speedrun.com/user/rewrkk',
 'name-style': {'style': 'gradient',
  'color-from': {'light': '#E762B5', 'dark': '#F772C5'},
  'color-to': {'light': '#900090', 'dark': '#A010A0'}},
 'role': 'user',
 'signup': '2023-07-15T07:37:31Z',
 'location': {'country': {'code': 'ph',
   'names': {'international': 'Philippines', 'japanese': None}}},
 'twitch': None,
 'hitbox': None,
 'youtube': {'uri': 'https://www.youtube.com/user/reWRKK'},
 'twitter': None,
 'speedrunslive': None,
 'assets': {'icon': {'uri': 'https://www.speedrun.com/static/user/j59wdvnx/icon?v=f9a6aff'},
  'supporterIcon': None,
  'image': {'uri': 'https://www.speedrun.com/static/user/j59wdvnx/image?v=2f81454'}},
 'links': [{'rel': 'self',
   'uri': 'https://www.speedrun.com/api/v1/users/j59wdvnx'},
  {'rel': 'runs', 'uri': 'https://www.speedrun.com/api/v1/runs?user=j59wdvnx'},
  

In [ ]:
api.get_user("j59wdvnx").data

https://www.speedrun.com/api/v1/users/j59wdvnx


{'id': 'j59wdvnx',
 'names': {'international': 'rewrkk', 'japanese': None},
 'supporterAnimation': False,
 'pronouns': 'Any/All',
 'weblink': 'https://www.speedrun.com/user/rewrkk',
 'name-style': {'style': 'gradient',
  'color-from': {'light': '#E762B5', 'dark': '#F772C5'},
  'color-to': {'light': '#900090', 'dark': '#A010A0'}},
 'role': 'user',
 'signup': '2023-07-15T07:37:31Z',
 'location': {'country': {'code': 'ph',
   'names': {'international': 'Philippines', 'japanese': None}}},
 'twitch': None,
 'hitbox': None,
 'youtube': {'uri': 'https://www.youtube.com/user/reWRKK'},
 'twitter': None,
 'speedrunslive': None,
 'assets': {'icon': {'uri': 'https://www.speedrun.com/static/user/j59wdvnx/icon?v=f9a6aff'},
  'supporterIcon': None,
  'image': {'uri': 'https://www.speedrun.com/static/user/j59wdvnx/image?v=2f81454'}},
 'links': [{'rel': 'self',
   'uri': 'https://www.speedrun.com/api/v1/users/j59wdvnx'},
  {'rel': 'runs', 'uri': 'https://www.speedrun.com/api/v1/runs?user=j59wdvnx'},
  

In [ ]:
api.get_game(dt.Game, id="v1pxjz68")

TypeError: SpeedrunCom.get_game() got multiple values for argument 'id'